cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+      debug::print(handle,
+                   matrix_partition.get_offsets() +
+                     (vertex_id - graph_view.get_local_adj_matrix_partition_row_first(i)),
+                   matrix_partition.get_offsets() +
+                     (vertex_id - graph_view.get_local_adj_matrix_partition_row_first(i)) + 2,
+                   offset_label,
+                   " ");
+    }
+  }
+}
+template <typename GraphViewType>
+void is_offset_sorted(raft::handle_t const& handle, GraphViewType graph_view, std::string str = "")
+{
+  using partition_t = cugraph::matrix_partition_device_view_t<typename GraphViewType::vertex_type,
+                                                              typename GraphViewType::edge_type,
+                                                              typename GraphViewType::weight_type,
+                                                              GraphViewType::is_multi_gpu>;
+  for (size_t i = 0; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+    auto matrix_partition = partition_t(graph_view.get_matrix_partition_view(i));
+    bool sorted =
+      thrust::is_sorted(handle.get_thrust_policy(),
+                        matrix_partition.get_offsets(),
+                        matrix_partition.get_offsets() + matrix_partition.get_major_size() + 1);
+    auto segment_offsets = graph_view.get_local_adj_matrix_partition_segment_offsets(i);
+    auto use_dcs         = segment_offsets
+                             ? ((*segment_offsets).size() >
+                        (cugraph::detail::num_sparse_segments_per_vertex_partition + 1))
+                             : false;
+
+    point(handle, "is_sorted ", static_cast<int>(sorted), i, use_dcs);
+  }
+}
+}  // namespace debug
+#endif
+namespace cugraph {
+
+namespace detail {
+
+/**
+ * @brief Calculate local out degrees of the sources belonging to the adjacency matrices
+ * stored on each gpu
+ *
+ * Iterate through partitions and store their local degrees
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return A single vector containing the local out degrees of the sources belong to the adjacency
+ * matrices
+ */
+template <typename GraphViewType>
+rmm::device_uvector<typename GraphViewType::edge_type> calculate_local_degrees(
+  raft::handle_t const& handle, GraphViewType const& graph_view)
+{
+  static_assert(GraphViewType::is_adj_matrix_transposed == false);
+  using vertex_t = typename GraphViewType::vertex_type;
+  using edge_t   = typename GraphViewType::edge_type;
+  using weight_t = typename GraphViewType::weight_type;
+
+  rmm::device_uvector<edge_t> local_degrees(
+    graph_view.get_number_of_local_adj_matrix_partition_rows(), handle.get_stream());
```

Comment:
Better be `GraphViewTypes::is_adj_matrix_transposed ? graph_view.get_number_of_local_adj_matrix_partition_cols() : graph_view.get_number_of_local_adj_matrix_partition_rows();` for better re-useability.